README.md
# PlagCheck ✅

> Moss Results scraper with powerful insights & analysis 💡

![PyPI](https://img.shields.io/pypi/v/plagcheck?color=blue)
[![Build Status](https://travis-ci.org/codeclassroom/PlagCheck.svg?branch=master)](https://travis-ci.org/codeclassroom/PlagCheck)
![PyPI - Python Version](https://img.shields.io/pypi/pyversions/plagcheck)
[![Documentation Status](https://readthedocs.org/projects/plagcheck/badge/?version=latest)](https://plagcheck.readthedocs.io/en/latest/?badge=latest)
![PyPI - License](https://img.shields.io/pypi/l/plagcheck?color=orange)
![PyPI - Downloads](https://img.shields.io/pypi/dm/plagcheck?color=blue)


## Installation

Install using `pip` from PyPI

```bash
pip install plagcheck
```

or directly from GitHub if you cannot wait to test new features

```bash
pip install git+https://github.com/codeclassroom/PlagCheck.git
```

## Usage

```python

"""Usage example"""
import os
import pprint
from plagcheck.plagcheck import check, insights, share_scores

from dotenv import load_dotenv
load_dotenv()

language = "java"
userid = os.environ["USER_ID"]


moss = check(language, userid)

moss.addFilesByWildCard("testfiles/test_java*.java")

# or moss.addFile("testfiles/test_python.py")

moss.submit()

print(moss.getHomePage())

result = moss.getResults()

pprint.pprint(result)

# print potential distributor-culprit relationships
pprint.pprint(insights(result))
# print frequency of each shared solution
pprint.pprint(share_scores(result))

```

## Documentation

> [PlagCheck Documentation](https://plagcheck.readthedocs.io/en/latest/)


## Development

##### Prerequisites
- Python 3.6+
- virtualenv

1. Create virtual environment.
```bash
virtualenv -p python3 venv && cd venv && source bin/activate
```
2. Clone the repository.
```bash
git https://github.com/codeclassroom/PlagCheck.git
```
3. Install Dependencies.
```bash
pip install -r requirements-dev.txt
```
4. Run tests.
```bash
pytest plagcheck
```
5. Lint the project with
```bash
flake8 plagcheck --max-line-length=88 --ignore=F401
black --check --diff plagcheck
```

## 📝 Changelog

See the [CHANGELOG.md](CHANGELOG.md) file for details.


## Author

👥 **Bhupesh Varshney**

- Twitter: [@bhupeshimself](https://twitter.com/bhupeshimself)
- DEV: [bhupesh](https://dev.to/bhupesh)

[![forthebadge](https://forthebadge.com/images/badges/built-with-love.svg)](https://forthebadge.com)

## 📜 License

This project is licensed under the MIT License. See the [LICENSE](LICENSE) file for details.

## 👋 Contributing

Please read the [CONTRIBUTING](CONTRIBUTING.md) guidelines for the process of submitting pull requests to us.


## Contributors ✨

Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/docs/en/emoji-key)):

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore -->
<table>
  <tr>
    <td align="center"><a href="https://github.com/vhsw"><img src="https://avatars3.githubusercontent.com/u/7099976?v=4" width="100px;" alt="Alexey Dubrov"/><br /><sub><b>Alexey Dubrov</b></sub></a><br /><a href="https://github.com/codeclassroom/PlagCheck/commits?author=vhsw" title="Code">💻</a> <a href="https://github.com/codeclassroom/PlagCheck/issues?q=author%3Avhsw" title="Bug reports">🐛</a> <a href="https://github.com/codeclassroom/PlagCheck/commits?author=vhsw" title="Tests">⚠️</a></td>
  </tr>
</table>

<!-- ALL-CONTRIBUTORS-LIST:END -->

This project follows the [all-contributors](https://github.com/all-contributors/all-contributors) specification. Contributions of any kind welcome!


CHANGELOG.md
# Changelog


## [0.4] - March 10, 2020

### Changed [⚠️ Breaking Changes]
- `getShareScores` & `getInsights` have been decoupled from the check class, they now have to be imported separately.
- Minor changes in the `analyze.py` module.


## [0.3] - Jan 1, 2020

### Added

-  New module `analyze.py` for Moss Results analysis
- `getShareScores()` for returning frequency of shared files.
- `addFile()` for adding files.
- `addFilesByWildCard()` for submitting multiple files.
- Support for adding base code using `addBaseCode()`.

### Changed
- The plagcheck module is now more modularised. `check` is now a class.
- `__get_line_numbers()` now runs in a new thread.

### Removed
- `requests` as a dependency, network requests are now 50% faster.


## [0.2] - Nov 9, 2019
- Minor Improvements


## [0.1] - Nov 3, 2019
- Initial Release


test/plagcheck_test.py
"""Tests for the MOSS interface package for CodeClassroom"""
import unittest
from plagcheck import plagcheck


class TestPlugCheck(unittest.TestCase):
    """Test cases"""

    def test_check(self):
        """General test"""
        program_files = ["testfiles/test_python.py", "testfiles/test_python3.py"]
        language = "python"
        userid = "1"
        url, results = plagcheck.check(program_files, language, userid)
        self.assertIn("http", url)
        self.assertEqual(
            results,
            [
                {
                    "file1": "testfiles/test_python.py",
                    "file2": "testfiles/test_python3.py",
                    "percentage": 90,
                    "no_of_lines_matched": 3,
                    "lines_matched": [["1-3", "1-3"]],
                }
            ],
        )

    def test_perc_str_to_int(self):
        """Test string parsing"""
        result = plagcheck.perc_str_to_int("(42%)")
        self.assertEqual(result, 42)


testfiles/test_java.java
class FibonacciExample1{
	public static void main(String args[]){  
	 int n1=0,n2=1,n3,i,count=10;  
	 System.out.print(n1+" "+n2);//printing 0 and 1  
	  
	 for(i=2;i<count;++i) {//loop starts from 2 because 0 and 1 are already printed  
		  n3=n1+n2;  
		  System.out.print(" "+n3);  
		  n1=n2;  
		  n2=n3;  
	 }  
	}
}

testfiles/test_python.py
word = str(input())
print("Hello, " + word)
print("your are awsm!")

PlagCheck/check.py
"""The MOSS interface package for CodeClassroom"""
import re
import requests
import mosspy  # type: ignore
from bs4 import BeautifulSoup as bs  # type: ignore


def download_report(url: str):
    """Download whole report locally including code diff links"""
    mosspy.download_report(url, "submission/report/", connections=8)


def check(program_files: list, lang: str, user_id: str) -> tuple:
    """Check files for same lines"""
    moss = mosspy.Moss(user_id, lang)
    for item in program_files:
        moss.addFile(item)
    url = moss.send()
    results = __extract_info(url)
    return url, results


def __get_line_numbers(url: str) -> list:
    """Get Line Numbers which are same"""

    list_of_line_nos = []
    result_page = re.sub(r".html$", "-top.html", url)

    res = requests.get(result_page, headers={"User-Agent": "Mozilla/5.0"})
    html = bs(res.text, "lxml")
    table = html.find("table")
    # skip header
    for row in table.find_all("tr")[1:]:
        matched_lines = []
        for line_nos in row.find_all("td"):
            line_nos = line_nos.text.strip()
            if line_nos:
                matched_lines.append(line_nos)
        list_of_line_nos.append(matched_lines)
    return list_of_line_nos


def __extract_info(url: str) -> list:
    """Scrape the webpage for file names, percentage match etc."""
    results = []
    res = requests.get(url, headers={"User-Agent": "Mozilla/5.0"})
    html = bs(res.text, "lxml")
    table = html.find("table")
    # table looks like this
    # <TABLE>
    # <TR><TH>File 1<TH>File 2<TH>Lines Matched
    # <TR>
    #     <TD><A HREF="http://.../match0.html">testfiles/test_java2.java (92%)</A>
    #     <TD><A HREF="http://.../match0.html">testfiles/test_java3.java (46%)</A>
    #     <TD ALIGN=right>9
    # <TR>
    #     <TD><A HREF="http://.../match1.html">testfiles/test_java.java (87%)</A>
    #     <TD><A HREF="http://.../match1.html">testfiles/test_java3.java (45%)</A>
    #     <TD ALIGN=right>8
    # </TABLE>
    # in order to parse this table we need:
    # skip header
    for row in table.find_all("tr")[1:]:
        # put each column in separate varible
        col1, col2, col3 = row.find_all("td")
        # get matched line ranges from reference from first column
        line_numbers = __get_line_numbers(col1.a.get("href"))
        # get total number of matched lines
        no_of_lines_matched = int(col3.text.strip())
        # get filename and raw percentage stirng from first column
        filename1, perc_str = col1.text.strip().split()
        # get filename from second column
        filename2, ________ = col2.text.strip().split()
        # parse raw percentage from "(45%)" to 45
        match = re.search(r"\((\d+)%\)$", perc_str)
        if match:
            perc = int(match.group(1))
        else:
            raise ValueError("cannot find percentage in table. See " + url)
        result_dict = dict(
            file1=filename1,
            file2=filename2,
            percentage=perc,
            no_of_lines_matched=no_of_lines_matched,
            lines_matched=line_numbers,
        )
        results.append(result_dict)
    return results


plagcheck/analyze.py
"""
analyze.py
====================================
The Analysis Module for Moss
(For info. on how this works contact varshneybhupesh@gmail.com)
"""


class Node:
    """A Single Submitted file"""

    def __init__(self, name):
        self.name = name
        self.tag = None
        self.links = []

    def pointTo(self):
        """Return all nodes a node points to"""
        return [link.name for link in self.links]


class Mgroups:
    """A Disconnected directed graph consisting all individual solutions"""

    linkCount = 0

    def __init__(self):
        self.nodes = []
        self.nodeCount = 0

    def relate(self, P1, P2, node1, node2):
        """Set a path between two file nodes"""
        node_obj_dict = {}

        for r in self.nodes:
            node_obj_dict[r.name] = r

        if node1 in node_obj_dict.keys() and node2 in node_obj_dict.keys():
            if P1 < P2:
                node_obj_dict[node1].links.append(node_obj_dict[node2])
            elif P1 > P2:
                node_obj_dict[node2].links.append(node_obj_dict[node1])
            else:
                node_obj_dict[node1].links.append(node_obj_dict[node2])
                node_obj_dict[node2].links.append(node_obj_dict[node1])

        Mgroups.linkCount += 1

    def __indegree(self, node: Node):
        indegree_count = 0
        for n in self.nodes:
            for link in n.pointTo():
                if link == node.name:
                    indegree_count += 1
        return indegree_count

    def __outdegree(self, node: Node):
        return len(node.links)

    def set_tags(self):
        """Assign appropriate tag to a Node"""
        for node in self.nodes:
            in_degree = self.__indegree(node)
            out_degree = self.__outdegree(node)

            if in_degree == 0 and out_degree > 0:
                node.tag = "D"
            elif in_degree > 0 and out_degree == 0:
                node.tag = "C"
            elif in_degree != 0 and out_degree != 0:
                node.tag = "DC"

    def createNodes(self, node_set: set):
        """Create multiple nodes at the same time"""
        for n in node_set:
            self.addNode(n)

    def addNode(self, name: str):
        """Add a single node to graph"""
        if name not in [r.name for r in self.nodes]:
            node = Node(name)
            self.nodes.append(node)
            self.nodeCount += 1
        return node

    def displayNodes(self):
        """Return all nodes in the graph"""
        return [r.name for r in self.nodes]

    def displayTags(self):
        """Display Nodes with their Tags"""
        for node in self.nodes:
            print("{}, tag = {}".format(node.name, node.tag))

    def d2c(self):
        """All Direct Distributor to Culprit paths"""
        paths = []
        for node in self.nodes:
            for link in node.links:
                if node.tag == "D" and link.tag == "C":
                    paths.append(tuple((node.name, link.name)))
        return paths

    def d2dc(self):
        """All Direct Distributor to potential Distributor-Culprit paths"""
        paths = []
        for node in self.nodes:
            for link in node.links:
                if node.tag == "D" and link.tag == "DC":
                    paths.append(tuple((node.name, link.name)))
        return paths

    def dc2c(self):
        """All potential Distributor-Culprit to direct Culprit paths"""
        paths = []
        for node in self.nodes:
            for link in node.links:
                if node.tag == "DC" and link.tag == "C":
                    paths.append(tuple((node.name, link.name)))
        return paths

    def __repr__(self):
        """Pretty prints the graph"""
        paths = ""
        for node in self.nodes:
            for link in node.links:
                paths += "{0} --> {1}\n".format(node.name, link.name)
        return paths


plagcheck/plagcheck.py
"""The MOSS interface package for CodeClassroom"""
import collections
import re
import urllib.request
from concurrent.futures import ThreadPoolExecutor
from typing import List, Tuple

import mosspy
from bs4 import BeautifulSoup as bs

from plagcheck.analyze import Mgroups

HEADERS = {"User-Agent": "Mozilla/5.0"}


class Result(dict):
    """Typing for moss results"""

    file1: str
    file2: str
    percentage: int
    no_of_lines_matched: int
    lines_matched: List[List[str]]


Results = List[Result]


def perc_str_to_int(string: str) -> int:
    """Convert string like "(42%)" to 42"""
    match = re.search(r"\((\d+)%\)$", string)
    if match:
        return int(match.group(1))
    raise ValueError("Cannot find percentage in table")


def request(url: str):
    """Request Webpage"""
    req = urllib.request.Request(url, headers=HEADERS)
    with urllib.request.urlopen(req) as response:
        req = response.read()

    return req.decode("utf-8")


def share_scores(moss_data: dict) -> dict:
    """Share Score Insights"""
    similar_code_files = []
    for result in moss_data:
        similar_code_files.append(result["file1"])
        similar_code_files.append(result["file2"])

    # frequency of files which are similar
    share_score = collections.Counter(similar_code_files)

    return dict(share_score)


def insights(moss_data: dict) -> dict:
    """Analysis for Moss"""
    mg = Mgroups()
    similar_code_files = set()
    insights = {}

    for r in moss_data:
        similar_code_files.add(r["file1"])
        similar_code_files.add(r["file2"])

    mg.createNodes(similar_code_files)

    for r in moss_data:
        mg.relate(
            r["percentage_file1"], r["percentage_file2"], r["file1"], r["file2"]
        )

    mg.set_tags()

    insights["DtoC Paths"] = mg.d2c()
    insights["DtoDC Paths"] = mg.d2dc()
    insights["DCtoC Paths"] = mg.dc2c()

    return insights


class check:
    """
    Args:
        - Program Files (list)
        - Language (str)
        - Moss User ID (str)
    """

    def __init__(self, lang: str, user_id: str):

        self.__user_id = user_id
        languages = mosspy.Moss.languages
        if lang not in languages:
            raise ValueError(f"{lang} is not a supported language {languages}")
        self.lang = lang
        self.__moss = mosspy.Moss(self.__user_id, self.lang)

    def __extract_info(self) -> Results:
        """Scrape the webpage for file names, percentage match etc."""
        results: Results = []

        response = request(self.home_url)

        html = bs(response, "lxml")
        table = html.find("table")
        for row in table.find_all("tr")[1:]:
            col1, col2, col3 = row.find_all("td")
            filename1, perc1 = col1.text.strip().split()
            filename2, perc2 = col2.text.strip().split()

            with ThreadPoolExecutor() as executor:
                future = executor.submit(self.__get_line_numbers, col1.a.get("href"))
                lines = future.result()

            result_dict = Result(
                file1=filename1,
                file2=filename2,
                percentage_file1=perc_str_to_int(perc1),
                percentage_file2=perc_str_to_int(perc2),
                no_of_lines_matched=int(col3.text.strip()),
                lines_matched=lines,
            )
            results.append(result_dict)
        return results

    def __get_line_numbers(self, url: str) -> List[List[str]]:
        """Get Line Numbers which are same"""
        list_of_line_nos: List[List[str]] = []
        result_page = re.sub(r".html$", "-top.html", url)

        response = request(result_page)

        html = bs(response, "lxml")
        table = html.find("table")
        for row in table.find_all("tr")[1:]:
            matched_lines: List[str] = []
            for col in row.find_all("td"):
                line_nos: str = col.text.strip()
                if line_nos:
                    matched_lines.append(line_nos)
            list_of_line_nos.append(matched_lines)
        return list_of_line_nos

    def addFilesByWildCard(self, files):
        """Add multiple files"""
        self.__moss.addFilesByWildcard(files)

    def addFile(self, file):
        """Add a single file for submission"""
        self.__moss.addFile(file)

    def addBaseCode(self, base_file: str):
        """Add basefile"""
        self.__moss.addBaseFile(base_file)

    def submit(self):
        """Submit files to the Moss Server"""
        url = self.__moss.send()

        self.home_url = url
        self.moss_results = self.__extract_info()

    def getHomePage(self):
        """Return Moss Results HomePage URL"""
        return self.home_url

    def getResults(self) -> Tuple[str, Results]:
        """Return the result as a list of dictionary"""

        return self.moss_results
